src/common-libs/auth/permission-decorator.ts

```diff
@@ -0,0 +1,27 @@
+import { actionIsPermitted } from '@unocha/hpc-api-core/src/auth';
+import { RequiredPermissionsCondition } from '@unocha/hpc-api-core/src/auth/permissions';
+import { Context } from '@unocha/hpc-api-core/src/lib/context';
+import { ForbiddenError } from '@unocha/hpc-api-core/src/util/error';
+import { createMethodDecorator, ResolverData } from 'type-graphql';
+
+type RequiredPermissions = (
+  resolverData: ResolverData<Context>
+) => Promise<RequiredPermissionsCondition<never>>;
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export function Permission(
+  requiredPermissions: RequiredPermissions
+): MethodDecorator {
+  return createMethodDecorator(
+    async (resolverData: ResolverData<Context>, next) => {
+      const permissions = await requiredPermissions(resolverData);
+      const { context } = resolverData;
+
+      if (!(await actionIsPermitted(permissions, context))) {
+        throw new ForbiddenError('No permission to perform this action');
+      }
+
+      return next();
+    }
+  );
+}
```

Comment:
The commit message for 83852014da82ce19462d617698dc21eca394a678 seems to misalign with the actual implementation of `@Permission`, which instead seems to take a function and return the required permissions, which I like much more than what the commit describes (and I think is what I suggested when Joseph originally implemented this).

I would perhaps extend this to also support simply taking a `RequiredPermissionsCondition` directly too, and checking whether the given `requiredPermissions` is a function, and treating appropriately, which would reduce the code needed when using this decorator in simple permissions such as global permissions only.

Reply:
> "Here you only use global permissions, don't wrap inside the function" 😄

Sounds like a very reasonable review comment haha ;) Could maybe even make a linting rule for it! :-D

Reply:
Right. Why keep simple interface, when we can make it possible to use it multiple ways, and then restrict in some cases :D